pyomo/contrib/pynumero/sparse/mpi_block_vector.py

```diff
@@ -0,0 +1,1291 @@
+        elif isinstance(other, BlockVector):
+            assert self.nblocks == other.nblocks, \
+                'Number of blocks mismatch: {} != {}'.format(self.nblocks, other.nblocks)
+            return self.dot(other.toMPIBlockVector(self.rank_ownership, self.mpi_comm))
+        elif isinstance(other, np.ndarray):
+            assert self.shape == other.shape, 'Dimension mismatch: {} != {}'.format(self.shape, other.shape)
+            other_bv = self.copy_structure()
+            other_bv.copyfrom(other)
+            return self.dot(other_bv)
+        else:
+            raise NotImplementedError('Operation not supported by MPIBlockVector')
+
+    @staticmethod
+    def _serialize_structure(block_vector):
+        """
+        Parameters
+        ----------
+        block_vector: BlockVector
+
+        Returns
+        -------
+        list
+        """
+        serialized_structure = list()
+        for ndx in range(block_vector.nblocks):
+            blk = block_vector.get_block(ndx)
+            if isinstance(blk, BlockVector):
+                serialized_structure.append(-1)
+                serialized_structure.append(blk.nblocks)
+                serialized_structure.extend(MPIBlockVector._serialize_structure(blk))
+            elif isinstance(blk, MPIBlockVector):
+                raise NotImplementedError('Operation not supported for MPIBlockVectors containing other MPIBlockVectors')
+            elif isinstance(blk, np.ndarray):
+                serialized_structure.append(-2)
+                serialized_structure.append(blk.size)
+            else:
+                raise NotImplementedError('Unrecognized input.')
+        return serialized_structure
+
+    @staticmethod
+    def _create_from_serialized_structure(serialized_structure, structure_ndx, result):
+        """
+        Parameters
+        ----------
+        serialized_structure: np.ndarray
+        structure_ndx: int
+        result: BlockVector
+
+        Returns
+        -------
+        structure_ndx: int
+        """
+        for ndx in range(result.nblocks):
+            if serialized_structure[structure_ndx] == -1:
+                structure_ndx += 1
+                result.set_block(ndx, BlockVector(serialized_structure[structure_ndx]))
+                structure_ndx += 1
+                structure_ndx = MPIBlockVector._create_from_serialized_structure(serialized_structure,
+                                                                                 structure_ndx,
+                                                                                 result.get_block(ndx))
+            elif serialized_structure[structure_ndx] == -2:
+                structure_ndx += 1
+                result.set_block(ndx, np.zeros(serialized_structure[structure_ndx]))
+                structure_ndx += 1
+            else:
+                raise ValueError('Unrecognized structure')
+        return structure_ndx
+
+    def make_local_structure_copy(self):
```

Comment:
Did we have this functionality on MPIBlockMatrix?

Reply:
No. I added this to MPIBlockVector to generalize the MPIBlockMatrix-MPIBlockVector dot product. We don't support MPIBlockMatrix*MPIBlockMatrix, so I didn't need this functionality for MPIBlockMatrix.